keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2488 @@
+spec:
+  allowed: ["a", "b", "c"]
+  banned: ["x", "y", "z"]
+  xyz:
+    fuzzFactor: 0.8
+    reticulate: true
+```
+
+Note that for policy configurations that require no parameterization, only the
+`PolicyConfiguration` is needed.
+
+Pros:
+
+- Matching criteria is fully defined and validated in a builtin type.
+- The configuration data that the policy definition is treated as an opaque
+  payload from an apiserver perspective, it passes it to the policy definition
+  so that CEL expressions can access it, but there is no code in the apiserver
+  that cares about the structure of the configuration CRD.
+
+Cons:
+
+- Requires two resources for each configuration of a policy
+  - Cannot atomically change the match rules and configuration of a policy configuration
+
+Alternative: OpenAPIv3 `$ref` in CRDs
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.rules.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                maxReplicas:
+                  type: int
+                match:
+                  $ref: "#/components/schemas/matchrules"
+  scope: Cluster
+  names:
+    kind: ReplicaLimit
+```
+
+Pros:
+
+- Match rule schema is owned by Kubernetes, so as it evolves, CRDs automatically
+  pick up changes
+
+Cons:
+
+- CRDs to not yet support OpenAPIv3 `$ref`s, so support would need to be added,
+  presumably this is a separate KEP
+
+
+Alternative: Duck Typed CRDs
+
+This is the alternative shown in the initial examples of this KEP.
+
+The idea of this alternative is to use duck typing: If a CRD author includes a
+`spec.match` schema in their CRD conforming to the schema expected (described
+below) and then for the apiserver access the data and treats it as match
+criteria.
+
+If the schema for `spec.match` it is malformed, errors are reported in the
+status of the `ValidatingAdmissionPolicy`.
+
+A partial `spec.match` schema (subset of the full schema) is okay so long as
+only optional fields are omitted. But any unrecognized field in the `spec.match`
+would not be allowed. 
+
+Pros:
+
+- A single resource is used to configure both the match criteria and
+  configuration params of a policy.
+- Policy authors can choose to only provide the subset of `spec.match` schema
+  that apply to the configuration of the policy they have defined. E.g. a policy
+  that is already constrained to pods by the policy definition does not
+  need to expose the match criteria for GVR matching.
+
+Cons:
+
+- API server must check for a wide range of error conditions and define how
+  exactly it handles each of them.
+- The way CRDs expose Match criteria is entirely implicit.
+- If the `spec.match` schema is incorrectly defined, CRD author might not
```

Comment:
What if you allowed an annotation on a CRD manifest that meant: "inject the correct .spec.match during admission and keep it up to date with a controller"?

Reply:
Oh nice. I've added this to the alternative proposal.